codec-dns/src/main/java/io/netty/handler/codec/dns/AbstractDnsMessage.java

```diff
@@ -0,0 +1,462 @@
+        recordList.add(record);
+    }
+
+    @Override
+    public DnsMessage addRecord(DnsSection section, int index, DnsRecord record) {
+        addRecord(sectionOrdinal(section), index, record);
+        return this;
+    }
+
+    private void addRecord(int section, int index, DnsRecord record) {
+        checkQuestion(section, record);
+
+        final Object records = sectionAt(section);
+        if (records == null) {
+            if (index != 0) {
+                throw new IndexOutOfBoundsException(String.valueOf(index));
+            }
+
+            setSection(section, record);
+            return;
+        }
+
+        if (records instanceof DnsRecord) {
+            final List<DnsRecord> recordList;
+            if (index == 0) {
+                recordList = newRecordList();
+                recordList.add(record);
+                recordList.add(castRecord(records));
+            } else if (index == 1) {
+                recordList = newRecordList();
+                recordList.add(castRecord(records));
+                recordList.add(record);
+            } else {
+                throw new IndexOutOfBoundsException(String.valueOf(index));
+            }
+            setSection(section, recordList);
+            return;
+        }
+
+        @SuppressWarnings("unchecked")
+        final List<DnsRecord> recordList = (List<DnsRecord>) records;
+        recordList.add(index, record);
+    }
+
+    @Override
+    public <T extends DnsRecord> T removeRecord(DnsSection section, int index) {
+        return removeRecord(sectionOrdinal(section), index);
+    }
+
+    private <T extends DnsRecord> T removeRecord(int section, int index) {
+        final Object records = sectionAt(section);
+        if (records == null) {
+            throw new IndexOutOfBoundsException(String.valueOf(index));
+        }
+
+        if (records instanceof DnsRecord) {
+            if (index != 0) {
+                throw new IndexOutOfBoundsException(String.valueOf(index));
+            }
+
+            T record = castRecord(records);
+            setSection(section, null);
+            return record;
+        }
+
+        @SuppressWarnings("unchecked")
+        final List<DnsRecord> recordList = (List<DnsRecord>) records;
+        return castRecord(recordList.remove(index));
+    }
+
+    @Override
+    public DnsMessage clear(DnsSection section) {
+        clear(sectionOrdinal(section));
+        return this;
+    }
+
+    @Override
+    public DnsMessage clear() {
+        for (int i = 0; i < SECTION_COUNT; i ++) {
+            clear(i);
+        }
+        return this;
+    }
+
+    private void clear(int section) {
+        final Object recordOrList = sectionAt(section);
+        setSection(section, null);
+        if (recordOrList != null) {
+            if (recordOrList instanceof ReferenceCounted) {
+                ((ReferenceCounted) recordOrList).release();
+            } else if (recordOrList instanceof List) {
```

Comment:
Mayne just eliminate the above if check and just do an "else" where you call ?ReferenceCountUtil.release(...)`